### PR TITLE
MGMT-8748: detect overlapping networks in SNO

### PR DESCRIPTION
**Pull request by slaviered**
# Assisted Pull Request

## Description

Fix current validation logic so it detects overlapping
Cluster and Machine Networks.

## List all the issues related to this PR

- [ ] New Feature
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/network/cidr_validations.go [hunk 1]
```diff
@@ -35,7 +27,11 @@ func VerifyCIDRsNotOverlap(aCidrStr, bCidrStr string) error {
 	if err != nil {
 		return err
 	}
-	return verifyCIDRsNotOverlap(acidr, bcidr)
+	//overlapping occur if one of the CIDRs is a subset of the other
```
Comment (ori-amizur) [inline internal/network/cidr_validations.go:30]: Why is this change needed ?

File: internal/cluster/validator.go [hunk 2]
```diff
@@ -498,31 +498,44 @@ func (v *clusterValidator) printIsDNSDomainDefined(context *clusterPreprocessCon
...
+	}
 
 	// TODO MGMT-7587: Support any number of subnets
 	// Assumes that the number of cluster networks equal to the number of service networks
-	for index := range c.cluster.ClusterNetworks {
-		machineNetworkCidr := ""
-		if len(c.cluster.MachineNetworks) > index {
-			machineNetworkCidr = string(c.cluster.MachineNetworks[index].Cidr)
-		}
+	if len(c.cluster.ClusterNetworks) != len(c.cluster.ServiceNetworks) {
```
Comment (ori-amizur) [inline internal/cluster/validator.go:533]: Not sure it needs to be checked here.  Maybe another validation ?

File: internal/cluster/validator.go [hunk 3]
```diff
@@ -532,20 +545,8 @@ func (v *clusterValidator) printNoCidrsOverlapping(c *clusterPreprocessContext,
...
-				machineNetworkCidr = string(c.cluster.MachineNetworks[index].Cidr)
-			}
-
-			if err := network.VerifyClusterCIDRsNotOverlap(machineNetworkCidr,
-				string(c.cluster.ClusterNetworks[index].Cidr),
-				string(c.cluster.ServiceNetworks[index].Cidr),
-				swag.BoolValue(c.cluster.UserManagedNetworking)); err != nil {
-				return fmt.Sprintf("CIDRS Overlapping: %s.", err.Error())
-			}
+		if err := checkCidrsOverlapping(c.cluster); err != nil {
```
Comment (ori-amizur) [inline internal/cluster/validator.go:548]: Why check again on ValidationFailure ?

Comment (slaviered) [reply internal/cluster/validator.go:548]: It's the code the already were there. I just put it into common function because fixes in the original flow did not propagate to this place. I guess that the original intent was to have different error messages in case of a failure

File: internal/cluster/validator.go (hunk 2 above)
Comment (slaviered) [reply internal/cluster/validator.go:533]: I return validation error. This is just formulating the pre-assumption for this very validation in the code

File: internal/network/cidr_validations.go (hunk 1 above)
Comment (slaviered) [reply internal/network/cidr_validations.go:30]: Why having so many sub functions with similar names? The 2 liners were used just here so I merged them. I found it clearer to understand as a reader and follow the flow

File: internal/cluster/validator.go (hunk 2 above)
Comment (ori-amizur) [reply internal/cluster/validator.go:533]: https://github.com/openshift/assisted-service/blob/d7580b33923b77f158e96eece5d65590ce16680e/internal/cluster/validator.go#L24

File: internal/cluster/validator.go (hunk 3 above)
Comment (ori-amizur) [reply internal/cluster/validator.go:548]: In case you return ValidationError instead of ValidationFailure in other cases, then you will not have to rerun
